src/supernova/utils.rs

```diff
@@ -23,60 +22,99 @@ use crate::{
-) -> Result<(Vec<Boolean>, AllocatedRelaxedR1CSInstance<G>), SynthesisError> {
-  let mut a = a.iter().enumerate();
-  let mut selector: Vec<Boolean> = Vec::with_capacity(a.len());
-
-  let first_selected = alloc_num_equals_const(
-    cs.namespace(|| "check 0 equal bit".to_string()),
-    target_index,
-    G::Base::from(0),
-  )?;
-
-  selector.push(Boolean::Is(first_selected.clone()));
+  selector_vec: &[Boolean],
+) -> Result<AllocatedRelaxedR1CSInstance<G>, SynthesisError> {
+  assert_eq!(a.len(), selector_vec.len());
 
+  // Compare all instances in `a` to the first one
   let first = a
-    .next()
-    .ok_or_else(|| SynthesisError::IncompatibleLengthVector("empty vec length".to_string()))?
-    .1
-    .clone();
-
-  let initial_sum = Boolean::from(first_selected).lc(CS::one(), G::Base::ONE);
-
-  let (selected, selected_sum) = a.try_fold(
-    (first, initial_sum),
-    |(matched, mut selected_sum), (i, candidate)| {
-      let equal_bit = Boolean::Is(alloc_num_equals_const(
-        cs.namespace(|| format!("check {:?} equal bit", i)),
-        target_index,
-        G::Base::from(i as u64),
-      )?);
-
-      selector.push(equal_bit.clone());
-      selected_sum = selected_sum + &equal_bit.lc(CS::one(), G::Base::ONE);
+    .get(0)
+    .ok_or_else(|| SynthesisError::IncompatibleLengthVector("empty vec length".to_string()))?;
+
+  // Since `selector_vec` is correct, only one entry is 1.
+  // If selector_vec[0] is 1, then all `conditionally_select` will return `first`.
+  // Otherwise, the correct instance will be selected.
+  let selected = a
+    .iter()
+    .zip(selector_vec.iter())
+    .enumerate()
+    .skip(1)
+    .try_fold(first.clone(), |matched, (i, (candidate, equal_bit))| {
       let next_matched_allocated = conditionally_select_alloc_relaxed_r1cs(
         cs.namespace(|| format!("next_matched_allocated-{:?}", i)),
         candidate,
         &matched,
-        &equal_bit,
+        equal_bit,
       )?;
 
-      Ok::<(AllocatedRelaxedR1CSInstance<G>, LinearCombination<G::Base>), SynthesisError>((
-        next_matched_allocated,
-        selected_sum,
-      ))
-    },
-  )?;
-
-  cs.enforce(
-    || "exactly-one-selection",
-    |_| selected_sum,
-    |lc| lc + CS::one(),
-    |lc| lc + CS::one(),
-  );
-
-  Ok((selector, selected))
+      Ok::<AllocatedRelaxedR1CSInstance<G>, SynthesisError>(next_matched_allocated)
+    })?;
+
+  Ok(selected)
+}
+
+/// Compute a selector vector `s` of size `num_indices`, such that
+/// `s[i] == 1` if i == `target_index` and 0 otherwise.
+pub fn get_selector_vec_from_index<F: PrimeField, CS: ConstraintSystem<F>>(
+  mut cs: CS,
+  target_index: &AllocatedNum<F>,
+  num_indices: usize,
+) -> Result<Vec<Boolean>, SynthesisError> {
+  assert_ne!(num_indices, 0);
+
+  // Trivial case: return [1], enforce `target_index == 0`
+  if num_indices == 1 {
+    cs.enforce(
+      || "target_index = 0",
+      |lc| lc,
+      |lc| lc,
+      |lc| lc + target_index.get_variable(),
+    );
```

Comment:
I think this is redundant. The general case will handle this one correctly, saving a constraint — but more importantly keeping the logic simple.